src/navigator_data_ingest/base/api_client.py

```diff
@@ -0,0 +1,142 @@
+    """
+    # download the document
+    _LOGGER.info(f"Downloading document from '{source_url}'")
+    upload_result = DocumentUploadResult(
+        cloud_url=None,
+        md5_sum=None,
+        content_type=None,
+    )
+
+    try:
+        download_response = session.get(source_url, allow_redirects=True, timeout=5)
+        if download_response.status_code >= 300:
+            _LOGGER.error(
+                f"Downloading source document failed: {download_response.status_code} "
+                f"{download_response.text}"
+            )
+            return upload_result
+
+        content_type = download_response.headers["Content-Type"].split(";")[0]
+        # Update the result object with the detected content type
+        upload_result.content_type = content_type
+
+        # Decide what to do next based on content type
+        if content_type in MULTI_FILE_CONTENT_TYPES:
+            raise UnsupportedContentTypeError(content_type)
+
+        if content_type not in SUPPORTED_CONTENT_TYPES:
+            raise UnsupportedContentTypeError(content_type)
+
+        # Calculate the m5sum & update the result object with the calculated value
+        file_content = download_response.content
+        file_content_hash = hashlib.md5(file_content).hexdigest()
+        upload_result.md5_sum = file_content_hash
+
+        # s3 can only handle paths of up to 1024 bytes. To ensure we don't exceed that,
+        # we trim the filename if it's too long
+        file_suffix = content_type.split("/")[1]
+        filename_max_len = (
+            1024
+            - len(file_name_without_suffix)
+            - len(file_suffix)
+            - len(file_content_hash)
+            - len("_.")  # length of additional characters for joining path components
+        )
+        file_name_without_suffix_trimmed = file_name_without_suffix[:filename_max_len]
+        file_name = (
+            f"{file_name_without_suffix_trimmed}_{file_content_hash}.{file_suffix}"
+        )
+
+        _LOGGER.info(
+            f"Uploading supported single file document content from '{source_url}' "
+            f"to CDN s3 bucket with filename '{file_name}'"
+        )
+        cdn_url = _store_cached_document(document_bucket, file_name, file_content)
+        upload_result.cloud_url = cdn_url
+
+    except UnsupportedContentTypeError as e:
+        _LOGGER.warn(
+            f"Uploads for document at '{source_url}' could not be completed because "
+            f"the content type '{e.content_type}' is not currently supported."
+        )
+    finally:
+        # Always return an upload result, even if it's incomplete
+        # TODO: perhaps use the existence of an incomplete output in the future
+        return upload_result
+
+
+def _store_cached_document(
+    bucket: str,
+    name: str,
+    data: bytes,
+) -> str:
+    clean_name = name.lstrip("/")
+    output_file_location = S3Path(f"s3://{bucket}/{clean_name}")
```

Comment:
you can `S3Path("bucket", "folder", "file.txt") `

Reply:
I'll do this